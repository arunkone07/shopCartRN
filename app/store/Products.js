import { createSlice } from "@reduxjs/toolkit";

let lastId = 0;

const slice = createSlice({
  name: "products",
  initialState: [],
  reducers: {
    
    productAdded: (products, action) => {
      products.push({
        id: action.payload.id,
        title: action.payload.title,
        price: action.payload.price,
        image: action.payload.image,
        counter: 0
      });
    },

    productAddedToCart: (products, action) => {
      let index = -1
      products.forEach((product) => {
        if (product.id === action.payload.id) index = products.indexOf(product);
      });
      products[index].counter=1
    },

    productRemovedFromCart: (products, action) => {
      let index = -1
      products.forEach((product) => {
        if (product.id === action.payload.id) index = products.indexOf(product);
      });
      products[index].counter=0
    },

  },
});

export const { productAdded, productAddedToCart, productRemovedFromCart } = slice.actions;
export default slice.reducer;
