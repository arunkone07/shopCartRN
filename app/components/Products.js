import React, { useEffect } from "react";

import { FlatList, StyleSheet, SafeAreaView } from "react-native";
import { useSelector, useDispatch } from "react-redux";

import ProductCard from "./ProductCard";

import { productAdded, productAddedToCart } from "../store/Products";
import { productToCart } from "../store/Cart";

import productsApi from "../api/GetProducts";

function Products() {

  const dispatch = useDispatch();

  useEffect(() => {
    loadProducts();
  }, []);

  const loadProducts = async () => {
    const response = await productsApi.getProducts();
    response.data.forEach((product) => dispatch(productAdded(product)))
  };

  
  const allProducts = useSelector((state) => state.products)

  return (
    <SafeAreaView>

      <FlatList
        data={allProducts}
        keyExtractor={(product) => product.id.toString()}

        renderItem={({ item }) => (
          <ProductCard
            id={item.id}
            title={item.title}
            price={"Rs." + item.price}
            image={item.image}
            counter={item.counter}
            handleAdd={() => {
              dispatch(productToCart(item));
              dispatch(productAddedToCart(item))
            }}
          />
        )}

      />

      {/* can use below button for navigation in case of Stack Navigator 
        <Button
          title="Go to Cart"
          onPress={() => navigation.navigate("Cart", { addedProducts })}
        /> 
      */}

    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  screen: {
    paddingHorizontal: 20,
    backgroundColor: "yellow",
  },
});

export default Products;
