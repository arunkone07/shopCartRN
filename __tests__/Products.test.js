import React from 'react';
import { render, screen } from '@testing-library/react-native';
import Products from '../app/components/Products';


import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { useDispatch, useSelector } from 'react-redux';

import productsReducer, { productAdded, productAddedToCart } from "../app/store/Products";
import cartReducer from "../app/store/Cart";

describe('Products test', () => {
    const reactRedux = { useDispatch, useSelector }
    const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
    const useSelectorMock = jest.spyOn(reactRedux, "useSelector");

    it('snapshot test', () => {
        const initialState = { };
        const mockStore = configureStore();
        let store = mockStore(initialState);
        const products = render(
            <Provider store={store}>
                <Products />
            </Provider>
        );

        expect(products.toJSON()).toMatchSnapshot();
    });

    test('should return the initial state', () => {
        expect(productsReducer(undefined, { type: undefined })).toEqual([])
    });

    test('should handle a product being added to an empty list', () => {
        const previousState = []
        expect(productsReducer(previousState, productAdded({ "id":1,
        "title":"Product 1","price":"10","image":"https://shop.teamsg.in/wp-content/uploads/2021/04/Blazetech-yellow-small-2-4-scaled.jpg","isAdded":0}))).toEqual([{ 
            id: 1,
            title: "Product 1",
            price: "10",
            image: "https://shop.teamsg.in/wp-content/uploads/2021/04/Blazetech-yellow-small-2-4-scaled.jpg",
            counter: 0
        }
        ])
    });

    test('should handle a product being added to an existing list', () => {
        const previousState = [{ 
            id: 1,
            title: "Product 1",
            price: "10",
            image: "https://shop.teamsg.in/wp-content/uploads/2021/04/Blazetech-yellow-small-2-4-scaled.jpg",
            counter: 0
        }]
        expect(productsReducer(previousState, productAdded({"id":2,"title":"Product 2","price":"100","image":"https://rukminim1.flixcart.com/image/416/416/jf751u80/bat/g/g/u/1-1-2-short-handle-virat-kohli-grand-edition-tennis-cricket-bat-original-imaf3j2zhmdnays3.jpeg?q=70","isAdded":0}))).toEqual([
            { 
                id: 1,
                title: "Product 1",
                price: "10",
                image: "https://shop.teamsg.in/wp-content/uploads/2021/04/Blazetech-yellow-small-2-4-scaled.jpg",
                counter: 0
            },
            { 
                id: 2,
                title: "Product 2",
                price: "100",
                image: "https://rukminim1.flixcart.com/image/416/416/jf751u80/bat/g/g/u/1-1-2-short-handle-virat-kohli-grand-edition-tennis-cricket-bat-original-imaf3j2zhmdnays3.jpeg?q=70",
                counter: 0
            }
        ])
      
    });

});
