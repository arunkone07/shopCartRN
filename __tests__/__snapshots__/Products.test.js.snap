// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products test snapshot test 1`] = `
<RCTSafeAreaView>
  <RCTScrollView
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
